enclave-tls/src/tls_wrappers/openssl/negotiate.c

```diff
@@ -0,0 +1,116 @@
+/* Copyright (c) 2020-2021 Alibaba Cloud and Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <enclave-tls/log.h>
+#include <enclave-tls/err.h>
+#include <enclave-tls/tls_wrapper.h>
+#include "openssl.h"
+
+#ifndef SSL_SGX_WRAPPER
+extern int verify_certificate(int preverify, X509_STORE_CTX *store);
+#endif
+
+tls_wrapper_err_t openssl_internal_negotiate(tls_wrapper_ctx_t *ctx, unsigned long conf_flags,
+					     int fd, int (*verify)(int, X509_STORE_CTX *))
+{
+	openssl_ctx_t *ssl_ctx = ctx->tls_private;
+
+	/*
+	 * Set the verification mode.
+	 * Refer to https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_verify.html
+	 *
+	 * client: SSL_VERIFY_PEER
+	 * server: SSL_VERIFY_NONE
+	 * client+mutual: SSL_VERIFY_PEER
+	 * server+mutual: SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT
+	 */
+	if (verify) {
+		int mode = SSL_VERIFY_NONE;
+
+		if (!(conf_flags & ENCLAVE_TLS_CONF_FLAGS_SERVER))
+			mode |= SSL_VERIFY_PEER;
+		else if (conf_flags & ENCLAVE_TLS_CONF_FLAGS_MUTUAL)
+			mode |= SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
+
+		SSL_CTX_set_verify(ssl_ctx->sctx, mode, verify);
+	}
+
+	SSL *ssl = SSL_new(ssl_ctx->sctx);
+	if (!ssl)
+		return -TLS_WRAPPER_ERR_NO_MEM;
+
+	X509_STORE *cert_store = SSL_CTX_get_cert_store(ssl_ctx->sctx);
+	X509_STORE_set_ex_data(cert_store, 0, ctx);
```

Comment:
Use X509_STORE_get_ex_new_index() to allocate a new index instead of hard coding zero.
```C
int
X509_STORE_get_ex_new_index(long argl, void *argp, CRYPTO_EX_new *new_func, CRYPTO_EX_dup *dup_func, CRYPTO_EX_free *free_func);

int
X509_STORE_set_ex_data(X509_STORE *store, int idx, void *arg);
```